src/test/java/com/google/sps/servlets/ContestServletTest.java

```diff
@@ -327,9 +331,97 @@ public void oneElection_contestMissingInDatastore_testDoGet() throws IOException
+                + "\"source\":\"Voter Information Project\"}],\"referendums\":[]}");
+  }
+
+  @Test
+  public void oneElection_candidateMissingInDatastore_testDoGet() throws IOException {
+    Entity electionEntity = new Entity("Election");
+    Entity contestEntity = new Entity("Contest");
+
+    Entity firstCandidateEntity = new Entity("Candidate");
+    Entity secondCandidateEntity = new Entity("Candidate");
+
+    electionEntity.setPropertiesFrom(electionEntityOne);
+    contestEntity.setPropertiesFrom(contestEntityOne);
+
+    firstCandidateEntity.setPropertiesFrom(candidateEntityOne);
+    secondCandidateEntity.setPropertiesFrom(candidateEntityTwo);
+
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+    long contestId = ds.put(contestEntity).getId();
+    long firstCandidateId = ds.put(firstCandidateEntity).getId();
+
+    // Delete the Entity from Datastore to mimic what happens at the beginning
+    // of a call to doGet() in ElectionServlet.
+    long secondCandidateId = ds.put(secondCandidateEntity).getId();
+    Key entityKey = KeyFactory.createKey("Candidate", secondCandidateEntity.getKey().getId());
+    ds.delete(entityKey);
+
+    Collection<Long> candidateSet = (Collection<Long>) contestEntity.getProperty("candidates");
+    candidateSet.add(firstCandidateId);
+    candidateSet.add(secondCandidateId);
+    ds.put(contestEntity);
+
+    Collection<Long> contestSet = (Collection<Long>) electionEntity.getProperty("contests");
+    contestSet.add(contestId);
+    ds.put(electionEntity);
+
+    when(httpServletRequest.getParameter("electionId")).thenReturn("9999");
+    when(httpServletResponse.getWriter()).thenReturn(printWriter);
+
+    ContestsServlet contestServlet = new ContestsServlet();
+    contestServlet.doGet(httpServletRequest, httpServletResponse);
+
+    verify(printWriter)
+        .println(
+            "{\"contests\":[{\"name\":\"myFirstContest\",\"candidates\":[{\"name\":\"myFirstCandidate\","
+                + "\"partyAffiliation\":\"myParty\",\"campaignSite\":\"myWebsite\",\"platformDescription\":"
+                + "\"This is a cool candidate.\",\"channels\":{}}],"
+                + "\"description\":\"This contest is important.\",\"source\":\"Voter Information Project\"}],\"referendums\":[]}");
+  }
+
+  @Test
+  public void oneElection_referendumMissingInDatastore_testDoGet() throws IOException {
+    Entity electionEntity = new Entity("Election");
+    Entity firstReferendumEntity = new Entity("Referendum");
+    Entity secondReferendumEntity = new Entity("Referendum");
+
+    electionEntity.setPropertiesFrom(electionEntityOne);
+    firstReferendumEntity.setPropertiesFrom(referendumEntityOne);
+    secondReferendumEntity.setPropertiesFrom(referendumEntityTwo);
+
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+    long referendumOneId = ds.put(firstReferendumEntity).getId();
+
+    long referendumTwoId = ds.put(secondReferendumEntity).getId();
```

Comment:
Couldn't you just make up some random ID instead?

Reply:
I could, but then I run the risk of generating a random ID that is already assigned to an Entity in Datastore. The deletion imitates what happens in the actual Servlet code.